### PR TITLE
Notebook and a utility API for dealing with categorical variables

### PR DESCRIPTION
**Pull request by apaleyes**
*Issue #, if available:* #1 

*Description of changes:* I am happy to iterate on this notebook and related code further, making changes big and small, maybe even discarding my current approach towards the problem. But at this point I've rewritten it like 10 times, and need new opinions.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by apaleyes** [comment]
There've been some offline conversations going on, I'll summarize the outcome.

Current implementation suggest to encode categories as a single variable with values 1, 10, 100, 1000, etc. That can screw some models, so is probably a poor option.

Another approach is to represent each category as a separate discrete variable with values 0 and 1, or a continuous variable with values between 0 and 1, and then round optimization output somehow. This approach also has its drawbacks, cause it may confuse user, or it may result in suboptimal output.

All in all, I am going to do a bit of reading about how people deal with optimizing one-hot encoded values, and update the PR after this analysis.

**Comment by apaleyes** [comment]
This update is still incomplete and maybe rough on the edges, but i hope it displays the direction of where categorical is headed. Feedback is welcome while I work on

**Comment by apaleyes** [comment]
@javiergonzalezh that was more or less my plan for the next step towards completing this change

File: emukit/core/continuous_parameter.py [hunk 1]
```diff
@@ -2,12 +2,14 @@
 # SPDX-License-Identifier: Apache-2.0
 
 
-from typing import Union
+from typing import List, Union
```
Comment (marpulli) [inline emukit/core/continuous_parameter.py]: Is List used?

File: emukit/core/parameter.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    def dimension(self) -> int:
+        """
+        Gives the dimension of the parameter.
+        """
+        return 1
+
+    @property
+    def model_parameters(self) -> List:
+        """
+        Gives the list of single dimentional model parameters the parameter corresponds to.
```
Comment (marpulli) [inline emukit/core/parameter.py]:
```suggestion
        Gives the list of single dimensional model parameters the parameter corresponds to.
```

File: emukit/core/parameter.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        """
+        return [self]
+
+    def round(self, x) -> np.ndarray:
```
Comment (marpulli) [inline emukit/core/parameter.py]:
```suggestion
    def round(self, x: np.ndarray) -> np.ndarray:
```

File: emukit/core/parameter_space.py [hunk 4]
```diff
@@ -37,7 +39,17 @@ def get_parameter_by_name(self, name):
         for param in self.parameters:
             if param.name == name:
                 return param
-        raise ValueError('Parameter with name ' + name + 'not found.')
+        raise ValueError('Parameter with name ' + name + ' not found.')
+
+    def round(self, x):
```
Comment (marpulli) [inline emukit/core/parameter_space.py]: + Doc string
```suggestion
    def round(self, x: np.ndarray) -> np.ndarray:
```

File: notebooks/Emukit-Categorical-with-Tensorflow.ipynb [hunk 5]
```diff
@@ -0,0 +1,326 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Categorical variables: optimizing neural network with TensorFlow and Emukit"
```
Comment (marpulli) [inline notebooks/Emukit-Categorical-with-Tensorflow.ipynb]:
```suggestion
    "# Categorical variables: optimizing a neural network with TensorFlow and Emukit"
```

File: notebooks/Emukit-Categorical-with-Tensorflow.ipynb [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+    "      tf.keras.layers.Dense(10, activation=tf.nn.softmax)\n",
+    "    ])\n",
+    "    model.compile(optimizer=optimizer,\n",
+    "                  loss='sparse_categorical_crossentropy',\n",
+    "                  metrics=['accuracy'])\n",
+    "\n",
+    "    model.fit(x_train, y_train, epochs=1)\n",
+    "    loss, accuracy = model.evaluate(x_test, y_test)\n",
+    "    \n",
+    "    return accuracy"
```
Comment (marpulli) [inline notebooks/Emukit-Categorical-with-Tensorflow.ipynb]: I think we should be returning negative accuracy because emukit minimises? I'm not 100% sure though

File: emukit/core/categorical_parameter.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+        self._cont_params = []
+        for column_idx in range(self.encodings.shape[1]):
+            cont_param = ContinuousParameter(name + '_' + str(column_idx),
+                                             np.min(self.encodings[:, column_idx]),
+                                             np.max(self.encodings[:, column_idx]))
+            self._cont_params.append(cont_param)
+
+    @property
+    def encodings(self):
```
Comment (marpulli) [inline emukit/core/categorical_parameter.py]:
```suggestion
    def encodings(self) -> np.ndarray:
```

File: emukit/core/continuous_parameter.py (hunk 1 above)
Comment (apaleyes) [reply emukit/core/continuous_parameter.py]: Nope, removed

File: notebooks/Emukit-Categorical-with-Tensorflow.ipynb (hunk 6 above)
Comment (apaleyes) [reply notebooks/Emukit-Categorical-with-Tensorflow.ipynb]: You are correct, did that in emukit_friendly function

File: emukit/core/parameter_space.py (hunk 4 above)
Comment (apaleyes) [reply emukit/core/parameter_space.py]: Added docstring there and to the rest of undocumented methods in this class

File: emukit/core/optimization/acquisition_optimizer.py [hunk 8]
```diff
@@ -69,10 +69,15 @@ def f_df(x):
...
             _log.info("Starting gradient-free optimization of acquisition function {}".format(type(acquisition)))
-            return self.gpyopt_acquisition_optimizer.optimize(f, None, None)
+            x, f_min = self.gpyopt_acquisition_optimizer.optimize(f, None, None)
+
+        rounded_x = self.space.round(x)
+        rounded_f_min = acquisition.evaluate(rounded_x)
+
+        return rounded_x, rounded_f_min
 
     def _validate_context_parameters(self, context):
```
Comment (javiergonzalezh) [inline emukit/core/optimization/acquisition_optimizer.py]: can you add a small comment about what why this is needed?

File: integration_tests/emukit/bayesian_optimization/test_optimization_with_categorical_variables.py [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+from emukit.model_wrappers import GPyModelWrapper
+
+
+def test_categorical_variables():
+    np.random.seed(123)
+
+    def objective(x):
+        return np.array(np.sum(x, axis=1).reshape(-1, 1))
+
+    carol_spirits = ['past', 'present', 'yet to come']
```
Comment (javiergonzalezh) [inline integration_tests/emukit/bayesian_optimization/test_optimization_with_categorical_variables.py]: I like the categories here :-D!

File: emukit/core/optimization/acquisition_optimizer.py (hunk 8 above)
Comment (apaleyes) [reply emukit/core/optimization/acquisition_optimizer.py]: Sure. Although I've butchered diff of this PR with merging develop, so you'll have to take my word for it.

Comment (apaleyes) [reply emukit/core/optimization/acquisition_optimizer.py]: In fact I take it back, you can take a look at the diff for the last commit, which just adds the comment

File: emukit/core/encodings.py [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+        except ValueError:
+            raise ValueError("Unknown category {}" + str(category))
+
+        return self.encodings[index].tolist()
+
+
+class OneHotEncoding(Encoding):
+    def __init__(self, categories: List):
+        """
+        Initializes an instance of OneHotEncoding class
```
[the remark this answers is not in the document]
Comment (apaleyes) [reply emukit/core/encodings.py]: I have even better link to kaggle, less likely to disappear. Will use that.

File: emukit/core/encodings.py [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+        :returns: A rounded row.
+        """
+        idx = x_row.argmax()
+        row_rounded = np.zeros(x_row.shape)
+        row_rounded[idx] = 1
+        return row_rounded
+
+
+class OrdinalEncoding(Encoding):
+    def __init__(self, categories: List):
```
[the remark this answers is not in the document]
Comment (apaleyes) [reply emukit/core/encodings.py]: Will do
